Update prometheus/client_golang to 1.21.0-rc.0

bwplotka opened the pull request:
### What type of PR is this?

/kind feature
/hold

#### What this PR does / why we need it:

As per [new client_golang release process](https://github.com/prometheus/client_golang/blob/main/RELEASE.md) we try to understand better potential compatibility issues for every minor release, against a bigger users.

This particular release can be tricky due to client_golang dependency relaxing label name and metric name validation with the model.NameValidationScheme global value change, so we want to understand the consequences.

#### Special notes for your reviewer:

Technically this change permits new characters in the Prometheus metric and label names. cc @dashpole 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

bwplotka commented:
Any chances to trigger tests which is primary reason for this PR? (: 

/test

dims commented:
/test all

dashpole commented:
For future reference, to update a dependency: 

```sh
./hack/pin-dependency.sh github.com/prometheus/client_golang v1.21.0-rc.0
./hack/update-vendor.sh
```

bwplotka commented:
Did 

```
./hack/pin-dependency.sh github.com/prometheus/client_golang v1.21.0-rc.0
./hack/update-vendor.sh

```

dims commented:
/ok-to-test

dims commented:
/assign @dashpole

rexagod commented:
/assign @dashpole 
/triage accepted

dims commented:
@bwplotka need to touch up the `unwanted-dependencies.json` file. the instructions are in the logs of the failed ci job. looks like 5 lines can be removed now.

https://storage.googleapis.com/kubernetes-ci-logs/pr-logs/pull/129752/pull-kubernetes-dependencies/1882421034980741120/build-log.txt
```
2025/01/23 13:34:38 Status diff:
  []uint8(
  	"""
  	... // 21 identical lines
  	      "github.com/google/cadvisor",
  	      "github.com/json-iterator/go",
- 	      "github.com/prometheus/client_golang",
  	      "github.com/prometheus/common",
  	      "github.com/sirupsen/logrus",
  	... // 60 identical lines
  	      "github.com/grpc-ecosystem/go-grpc-middleware",
  	      "github.com/grpc-ecosystem/grpc-gateway",
- 	      "github.com/prometheus/client_golang",
  	      "go.etcd.io/etcd/api/v3",
  	      "go.etcd.io/etcd/client/v3",
  	... // 82 identical lines
  	      "github.com/opencontainers/runc",
  	      "k8s.io/mount-utils"
- 	    ],
- 	    "google.golang.org/appengine": [
- 	      "github.com/prometheus/client_golang"
  	    ],
  	    "google.golang.org/genproto": [
  	... // 33 identical lines
  	"""
  )
2025/01/23 13:34:38 Good news! Unwanted dependency "google.golang.org/appengine" is no longer referenced. Remove status.unwantedReferences["google.golang.org/appengine"] in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json to ensure it doesn't get reintroduced.
2025/01/23 13:34:38 Good news! Unwanted module "github.com/golang/protobuf" dropped the following dependants:
2025/01/23 13:34:38    github.com/prometheus/client_golang
2025/01/23 13:34:38 !!! Remove those from status.unwantedReferences["github.com/golang/protobuf"] in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json to ensure they don't get reintroduced.
2025/01/23 13:34:38 Good news! Unwanted module "github.com/davecgh/go-spew" dropped the following dependants:
2025/01/23 13:34:38    github.com/prometheus/client_golang
2025/01/23 13:34:38 !!! Remove those from status.unwantedReferences["github.com/davecgh/go-spew"] in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json to ensure they don't get reintroduced.
exit status 1
Removing /tmp/verify-vendor.sh.vn7rAV
```

bwplotka commented:
Yup, on it, thanks!

bwplotka committed:
Update prometheus/client_golang to 1.21.0-rc.0

As per new client_golang release process we try to understand better potential compatibility issues for every minor release, against a bigger users.

This particular release can be tricky due to client_golang dependency relaxing label name and metric name validation with the model.NameValidationScheme global value change, so we want to understand the consequences.

Special notes for your reviewer:

Technically this change permits new characters in the Prometheus metric and label names.

With this we also managed to remove the need for appeengine and locally go-spew, but we added zstd compression dep.

Signed-off-by: bwplotka <bwplotka@gmail.com>

bwplotka commented:
Done @dims @dashpole, thanks for help! Added tips for client_golang maintainers for the next PRs https://github.com/prometheus/client_golang/pull/1723

dashpole reviewed (approved):
/lgtm

dims commented:
/approve
/lgtm

/assign @liggitt

dims commented:
/hold do we want to wait till the final is cut?

rexagod commented:
`v1.22` is out.

dims commented:
fixed in https://github.com/kubernetes/kubernetes/pull/131204
